Reader.md

```diff
@@ -0,0 +1,176 @@
+# Intro
+
+Welcome to the series of videos about functional programming in Kotlin with Arrow. Arrow is a library that is packed with data types and type classes bringing typed FP to Kotlin. In this video, we're going to learn about the Reader type and what it's used for.
+
+# Slide 1
+```
+Reader
+
+Reader is a higher kinded wrapper around a function. This function will be ran at some point in the future, when we can provide a proper execution context for it.
+
+An execution context could be an input value, a service instance, or any other resource that the funtion could need to be ran.
+
+```
+
+# Slide 2
+```
+Reader
+
+In Arrow, it is actually defined in the terms of a Kleisli:
+
+typealias Reader<D, A> = ReaderT<IdHK, D, A>
+typealias ReaderT<F, D, A> = Kleisli<F, D, A>
+
+Reader is available in the arrow-data module under the import arrow.data.Reader
+
+// gradle
+compile ' io.arrow-kt:arrow-data:$arrowVersion'
+
+import arrow.data.Reader
+```
+
+# Slide 3
+```
+ReaderFun
+
+The function that Reader postpons is called `Computation` and has the form of (D) -> A, where D represents a Dependency. This computation is defined as an alias called `ReaderFun`.
+
+fun charExists(): ReaderFun<CharContext, Boolean> = { ctx ->
+    ctx.source.contains(ctx.searchChar)
+}
+```
+
+# Slide 4
+```
+Reader :: Instantiation
+
+Reader can be instantiated in three different ways.
+
+```
+
+# Slide 5
+```
+Reader :: using its constructor
+
+val readerInstance1: Reader<Int, String> = readerFun.reader()
```

Comment:
`readerFun` is used but never declared, perhaps in these examples is easier to just declare it directly `{ .. }.reader()` which is more explicit when showing the kind of function that is expected.

Reply:
Fixed, I forgot to update the name of the ReaderFun method in these examples. I thought on the inline instantiation but doing that is not clear to see the difference between the different methods. For instance:
```Reader({ctx: CharContext -> ctx.source.contains(ctx.searchChar)})```
vs
```Reader { ctx: CharContext -> ctx.source.contains(ctx.searchChar) }```
vs
```{ctx: CharContext -> ctx.source.contains(ctx.searchChar)}.reader()```

I see it a bit difficult to see especially for the 2 first use cases.

But I'd be happy to change it if you think it's clearer this way, no prob at all :)

Reply:
Makes sense but we can skip the first one all together since that is Kotlin specific and not Arrow that you can use parens around if you want.